intellij/src/saros/intellij/filesystem/IntellijFolderImplV2.java

```diff
@@ -0,0 +1,195 @@
+package saros.intellij.filesystem;
+
+import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.util.ThrowableComputable;
+import com.intellij.openapi.vfs.VirtualFile;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import saros.filesystem.IFile;
+import saros.filesystem.IFolder;
+import saros.filesystem.IPath;
+import saros.filesystem.IResource;
+import saros.intellij.project.filesystem.IntelliJPathImpl;
+import saros.intellij.runtime.FilesystemRunner;
+
+/** Intellij implementation of the Saros folder interface. */
+// TODO rename to IntellijFolder
+public class IntellijFolderImplV2 extends IntellijResourceImplV2 implements IFolder {
+  private static final Logger log = Logger.getLogger(IntellijFolderImplV2.class);
+
+  public IntellijFolderImplV2(
+      @NotNull IntellijReferencePointImpl referencePoint, @NotNull IPath relativePath) {
+
+    super(referencePoint, relativePath);
+  }
+
+  @Override
+  public boolean exists() {
+    VirtualFile virtualFile = getVirtualFile();
+
+    return existsInternal(virtualFile);
+  }
+
+  /**
+   * whether the given virtual file is not <code>null</code>, exists, and is a directory.
+   *
+   * @param virtualFile the virtual file to check
+   * @return whether the given virtual file is not <code>null</code>, exists, and is a directory
+   */
+  private static boolean existsInternal(@Nullable VirtualFile virtualFile) {
+    return virtualFile != null && virtualFile.exists() && virtualFile.isDirectory();
+  }
+
+  @Override
+  public boolean exists(@NotNull IPath path) {
+    return referencePoint.exists(relativePath.append(path));
+  }
+
+  @Override
+  @NotNull
+  public IResource[] members() throws IOException {
```

Comment:
Any plans on changing the return type (of the interface) to List? The conversions array, list, array in the code path are a bit ugly.

Reply:
Well, I would definitely prefer not to do it in this PR chain as this is defined by the core interface.

But yes, the conversions are ugly and the returned value is currently always converted back into a list by the caller anyways. Will adjust it separately.